File: 14-Stack-and-Queue-Part-II/python/09_maximum_of_minimums_for_every_window_size.py
class Solution:
    def maxOfMin(self, arr: List[int]) -> List[int]:
        n = len(arr)
        result = [0] * n
        left = [0] * n
        right = [0] * n
        stack = []

        for i in range(n):
            while stack and arr[stack[-1]] >= arr[i]:
                stack.pop()
            left[i] = -1 if not stack else stack[-1]
            stack.append(i)

        stack.clear()

        for i in range(n - 1, -1, -1):
            while stack and arr[stack[-1]] >= arr[i]:
                stack.pop()
            right[i] = n if not stack else stack[-1]
            stack.append(i)

        for i in range(n):
            window_size = right[i] - left[i] - 1
            idx = window_size - 1
            result[idx] = max(result[idx], arr[i])

        for i in range(n - 2, -1, -1):
            result[i] = max(result[i], result[i + 1])

        return result

File: 14-Stack-and-Queue-Part-II/cpp/09-maximum-of-minimums-for-every-window-size.cpp
class Solution {
public:
    vector<int> maxOfMin(vector<int>& arr) {
        int n = arr.size();
        vector<int> result(n, 0);
        vector<int> left(n), right(n);
        stack<int> st;

        for (int i = 0; i < n; i++) {
            while (!st.empty() && arr[st.top()] >= arr[i]) {
                st.pop();
            }
            left[i] = st.empty() ? -1 : st.top();
            st.push(i);
        }

        while (!st.empty()) st.pop();

        for (int i = n - 1; i >= 0; i--) {
            while (!st.empty() && arr[st.top()] >= arr[i]) {
                st.pop();
            }
            right[i] = st.empty() ? n : st.top();
            st.push(i);
        }

        for (int i = 0; i < n; i++) {
            int windowSize = right[i] - left[i] - 1;
            int idx = windowSize - 1;
            result[idx] = max(result[idx], arr[i]);
        }

        for (int i = n - 2; i >= 0; i--) {
            result[i] = max(result[i], result[i + 1]);
        }

        return result;
    }
};
